contentpal/src/test/java/org/dmfs/android/contentpal/utils/ContentOperationBuilderCountMatcher.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.android.contentpal.utils;
+
+import android.content.ContentProviderOperation;
+
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+
+import java.lang.reflect.Field;
+import java.util.Locale;
+
+
+/**
+ * A {@link ContentProviderOperation.Builder} {@link Matcher} which matches the operation type.
+ * <p>
+ * TODO: match Uri
+ *
+ * @author Marten Gajda
+ */
+public final class ContentOperationBuilderCountMatcher extends TypeSafeDiagnosingMatcher<ContentProviderOperation.Builder>
+{
+
+    private final Integer mExpectedCount;
+
+
+    public static ContentOperationBuilderCountMatcher expectingCount(int expectedCount)
+    {
+        return new ContentOperationBuilderCountMatcher(expectedCount);
+    }
+
+
+    public static ContentOperationBuilderCountMatcher expectingNoCount()
+    {
+        return new ContentOperationBuilderCountMatcher(null);
+    }
+
+
+    public ContentOperationBuilderCountMatcher(Integer expectedCount)
+    {
+        mExpectedCount = expectedCount;
+    }
+
+
+    @Override
+    protected boolean matchesSafely(ContentProviderOperation.Builder builder, Description mismatchDescription)
+    {
+        // create a builder an test the ContentValues
+        try
+        {
+            Field valuesField = ContentProviderOperation.Builder.class.getDeclaredField("mExpectedCount");
+            valuesField.setAccessible(true);
+
+            Integer expectedCount = (Integer) valuesField.get(builder);
+
+            if (mExpectedCount == null && expectedCount != null)
+            {
+                mismatchDescription.appendText(String.format(Locale.ENGLISH, "expects %d results", expectedCount));
+                return false;
+            }
+            if (mExpectedCount != null && expectedCount == null)
+            {
+                mismatchDescription.appendText("expects no specific number of results");
```

Comment:
Not sure about the correctness of the texts in this Matcher, so I just tried some tests, here are the messages I got:

```
Expected: (is a INSERT operation and expects 3 results and ((has value of key "account_type" is "type" and has value of key "account_name" is "name") and has 2 values))
but: expects 3 results expects no specific number of results


Expected: (is a UPDATE operation and expects 8 results and has 0 values)
but: expects 8 results expects 10 results


Expected: (is a UPDATE operation and expects no specific number of results and has 0 values)
but: expects no specific number of results expects 10 results
```
I guess the 'but' part should only contain the actual value description.

Reply:
I've noticed that too. Not sure why it also adds the "expected description". I'll have to debug the tests to find the problem.

Reply:
Looks like this is caused by an issue in the `AllOf` implementation in `org.hamcrest.core`, which contains:
```java
 @Override
    public boolean matches(Object o, Description mismatch) {
        for (Matcher<? super T> matcher : matchers) {
            if (!matcher.matches(o)) {
                mismatch.appendDescriptionOf(matcher).appendText(" ");
                matcher.describeMismatch(o, mismatch);
              return false;
            }
        }
        return true;
    }
```
it always appends the regular description of the matcher which fails. So maybe it's not intended to be used like I do!?

Reply:
Looks like this in intended: https://stackoverflow.com/questions/20566668/strange-allof-hamcrest-matcher-mismatch-description